File: src/search/utils/test_model_utils.cpp
/****************************************************\
 *
 * Copyright (C) 2019 All Rights Reserved
 * Last modified: 2026.09.18 18:08:50
 *
\****************************************************/

#include <gtest/gtest.h>
#define C10_USE_GLOG
#include <torch/torch.h>
#include <torch/script.h>
#include <glog/logging.h>
#include <safetensors.hh>
#include <file_utils.h>
#include "model_utils.h"

struct MyModel : torch::nn::Module {
  MyModel(std::int64_t hidden_size = 4096, std::int64_t intermediate_size = 11008) {
    fc1 = register_module("fc1", torch::nn::Linear(hidden_size, intermediate_size));
    fc2 = register_module("fc2", torch::nn::Linear(intermediate_size, hidden_size));
  }

  std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> forward_with_intermediates(
      const torch::Tensor & input) {
    auto fc1_output = fc1->forward(input);
    auto silu_output = torch::silu(fc1_output);
    auto output = fc2->forward(silu_output);
    return {fc1_output, silu_output, output};
  }

  torch::Tensor forward(const torch::Tensor & input) {
    return std::get<2>(forward_with_intermediates(input));
  }

  torch::nn::Linear fc1{nullptr};
  torch::nn::Linear fc2{nullptr};
};

TEST(Utils, SafeTensors) {
  std::string file_path = "model.safetensors";
  // Parse
  std::string warn;
  std::string err;
  safetensors::safetensors_t st;
  if (!safetensors::load_from_file(file_path, &st, &warn, &err)) {
    throw std::runtime_error("Failed to parse safetensors[" + err + "]");
  }
  if (!safetensors::validate_data_offsets(st, err)) {
    throw std::runtime_error("Invalid data_offsets[" + err + "]");
  }

  const std::vector<std::string> tensor_keys = st.tensors.keys();
  for (auto &key : tensor_keys) {
    LOG(INFO) << "Key[" << key << "]";
  }
}

TEST(Utils, ModelUtils) {
  std::string content;
  mycommon::file_read("model_utils.pt", content);
  torch::IValue ivalue = torch::jit::pickle_load(std::vector<char>(content.begin(), content.end()));
  if (!ivalue.isGenericDict()) {
    LOG(WARNING) << "Loaded data is not a dictionary!";
    ASSERT_EQ(true, false);
  }
	auto data = ivalue.toGenericDict();
	int hidden_size = data.at("hidden_size").toInt();
	int intermediate_size = data.at("intermediate_size").toInt();
	auto input = data.at("input").toTensor();
	auto ref_fc1_output = data.at("fc1_output").toTensor();
	auto ref_silu_output = data.at("silu_output").toTensor();
	auto ref_output = data.at("output").toTensor();
  LOG(INFO) << "hidden_size[" << hidden_size << "] intermediate_size[" << intermediate_size << "]";
  
  MyModel model(hidden_size, intermediate_size);
  load_module_from_safetensors(model, "model.safetensors");
  model.eval();
  //model.to(torch::kCUDA);
	
  torch::InferenceMode inference_mode;
  const auto [fc1_output, silu_output, output] = model.forward_with_intermediates(input);
  
  LOG(INFO) << format_tensor(input);
  LOG(INFO) << format_tensor(fc1_output);
  LOG(INFO) << format_tensor(ref_fc1_output);
  
  double rtol = 1e-5;
	double atol = 1e-5;
	//ASSERT_EQ(true, torch::allclose(fc1_output, ref_fc1_output, rtol, atol));
	//ASSERT_EQ(true, torch::allclose(silu_output, ref_silu_output, rtol, atol));
	//ASSERT_EQ(true, torch::allclose(output, ref_output, rtol, atol));
	ASSERT_EQ(true, check_close("fc1_output", fc1_output, ref_fc1_output, rtol, atol));
	ASSERT_EQ(true, check_close("silu_output", silu_output, ref_silu_output, rtol, atol));
	ASSERT_EQ(true, check_close("output", output, ref_output, rtol, atol));
}

int main(int argc, char *argv[]) {
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

/* vim: set expandtab nu ts=2 sw=2 sts=2: */

File: src/search/utils/test_model_utils.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import torch
from safetensors.torch import load_file, save_file

class MyModel(torch.nn.Module):
    """Python equivalent of the C++ MyModel used by the alignment test."""

    def __init__(self, hidden_size: int = 4096, intermediate_size: int = 11008):
        super().__init__()
        self.fc1 = torch.nn.Linear(hidden_size, intermediate_size)
        self.fc2 = torch.nn.Linear(intermediate_size, hidden_size)

    def forward_with_intermediates(
        self, x: torch.Tensor
    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
        fc1_output = self.fc1(x)
        silu_output = torch.nn.functional.silu(fc1_output)
        output = self.fc2(silu_output)
        return fc1_output, silu_output, output

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.forward_with_intermediates(x)[-1]

def initialize_deterministically(model: MyModel, seed: int) -> None:
    """Initialize all parameters from one CPU generator with a stable order."""

    generator = torch.Generator(device="cpu")
    generator.manual_seed(seed)
    with torch.no_grad():
        for parameter in model.parameters():
            parameter.copy_(
                torch.empty_like(parameter).uniform_(-0.05, 0.05, generator=generator)
            )

hidden_size = 4096
intermediate_size = 11008
batch_size = 3
seed = 20260915
root_dir = "../../../amd64/test/"

torch.set_num_threads(1)
torch.use_deterministic_algorithms(True)
model = MyModel(hidden_size, intermediate_size).eval()
initialize_deterministically(model, seed)
weights = {name: tensor.detach().cpu().contiguous() for name, tensor in model.state_dict().items()}
save_file(weights, root_dir + "model.safetensors")

input_generator = torch.Generator(device="cpu")
input_generator.manual_seed(seed + 1)
input_tensor = torch.randn(
    batch_size, hidden_size, generator=input_generator, dtype=torch.float32
)

with torch.inference_mode():
    fc1_output, silu_output, output = model.forward_with_intermediates(input_tensor)
torch.save({
  "hidden_size": hidden_size,
  "intermediate_size": intermediate_size,
  "batch_size": batch_size,
  "seed": seed,
  "input": input_tensor,
  "fc1_output": fc1_output,
  "silu_output": silu_output,
  "output": output
  }, root_dir + "model_utils.pt")

# vim: set expandtab nu ts=2 sw=2 sts=2:
